### PR TITLE
Feature/156400 persistence snapshots

### PR DESCRIPTION
**Pull request by ilya-martynov**

### CODE REVIEW

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

File: app/models/akka/DSLinkStateSnapshotter.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+      if (offeredSnapshot.responderBehaviorState != null) {
+        if (offeredSnapshot.responderBehaviorState.main != null)
+          receiveRecover(offeredSnapshot.responderBehaviorState.main)
+        if (offeredSnapshot.responderBehaviorState.additional != null)
+          receiveRecover(offeredSnapshot.responderBehaviorState.additional)
+      }
+
+    case RecoveryCompleted =>
+      log.debug("{}: recovery completed with persistenceId: '{}'", ownId, persistenceId)
```
Comment (dmitriyisaev) [inline app/models/akka/DSLinkStateSnapshotter.scala]: Info would be better

Comment (dmitriyisaev) [reply app/models/akka/DSLinkStateSnapshotter.scala]: info level i mean

File: app/models/akka/DSLinkStateSnapshotter.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      super.saveSnapshot(snapshot)
+    }
+  }
+}
+
+object DSLinkStateSnapshotter {
+  /**
+    * How many snapshots we want to keep after old ones removing.
+    */
+  val SNAPSHOTS_NUMBER_TO_KEEP = 3
```
Comment (dmitriyisaev) [inline app/models/akka/DSLinkStateSnapshotter.scala]: move 2 config?

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

File: app/models/akka/RequesterBehavior.scala [hunk 3]
```diff
@@ -30,8 +31,8 @@ trait RequesterBehavior { me: AbstractDSLinkActor with Meter =>
   ), "stateKeeper")
 
   // used by Close and Unsubscribe requests to retrieve the targets of previously used RID/SID
-  private val targetsByRid = collection.mutable.Map.empty[Int, String]
-  private val targetsBySid = collection.mutable.Map.empty[Int, PathAndQos]
+  private var targetsByRid = collection.mutable.Map.empty[Int, String]
+  private var targetsBySid = collection.mutable.Map.empty[Int, PathAndQos]
```
Comment (dmitriyisaev) [inline app/models/akka/RequesterBehavior.scala:36]: var and mutable map. Should we use val?

File: app/models/akka/responder/GroupCallRegistry.scala [hunk 4]
```diff
@@ -24,29 +24,57 @@ class GroupCallRecord {
 /**
  * Manages the call bindings for multi-recipient responses.
  */
-abstract class GroupCallRegistry(log: LoggingAdapter, ownId: String) {
+abstract class GroupCallRegistry(persistenceBehavior: PartOfPersistenceBehavior) {
 
-  private var bindings = Map.empty[Int, GroupCallRecord]
+  private var _bindings = Map.empty[Int, GroupCallRecord]
```
Comment (dmitriyisaev) [inline app/models/akka/responder/GroupCallRegistry.scala:29]: Doesn't really matter, but i think we can skip _ in ptivate fields.

Comment (eduardyuzlikeev) [reply app/models/akka/responder/GroupCallRegistry.scala:29]: IMHO it's a good idea since we use "_bindings_ as a backing field

Comment (dmitriyisaev) [reply app/models/akka/responder/GroupCallRegistry.scala:29]: ok

File: app/models/akka/RequesterBehavior.scala (hunk 3 above)
Comment (eduardyuzlikeev) [reply app/models/akka/RequesterBehavior.scala:36]: Nope. We replace this state in case of recovery at 139-140 lines.

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Pull request merged**